# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: parser/html/nsHtml5TreeBuilderCppSupplement.h
# Commit: 77388b0795b4
# Full Hash: 77388b0795b4f9a30171c922d6b649c52e5f6c87
# Author: Henri Sivonen <hsivonen@hsivonen.fi>
# Date: 2021-12-23 09:34:36
# Description:
#   Bug 1745142 - Communicate encoding commitment via speculative load queue. r=smaug
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D133996
# ==============================================================================

diff -r a388127095db -r 77388b0795b4 parser/html/nsHtml5TreeBuilderCppSupplement.h
--- a/parser/html/nsHtml5TreeBuilderCppSupplement.h	Wed Dec 22 15:47:08 2021 +0000
+++ b/parser/html/nsHtml5TreeBuilderCppSupplement.h	Wed Dec 22 15:54:49 2021 +0000
@@ -40,7 +40,8 @@
       mSpeculativeLoadStage(nullptr),
       mBroken(NS_OK),
       mCurrentHtmlScriptIsAsyncOrDefer(false),
-      mPreventScriptExecution(false)
+      mPreventScriptExecution(false),
+      mGenerateSpeculativeLoads(false)
 #ifdef DEBUG
       ,
       mActive(false)
@@ -50,7 +51,8 @@
 }
 
 nsHtml5TreeBuilder::nsHtml5TreeBuilder(nsAHtml5TreeOpSink* aOpSink,
-                                       nsHtml5TreeOpStage* aStage)
+                                       nsHtml5TreeOpStage* aStage,
+                                       bool aGenerateSpeculativeLoads)
     : mode(0),
       originalMode(0),
       framesetOk(false),
@@ -79,12 +81,15 @@
       mSpeculativeLoadStage(aStage),
       mBroken(NS_OK),
       mCurrentHtmlScriptIsAsyncOrDefer(false),
-      mPreventScriptExecution(false)
+      mPreventScriptExecution(false),
+      mGenerateSpeculativeLoads(aGenerateSpeculativeLoads)
 #ifdef DEBUG
       ,
       mActive(false)
 #endif
 {
+  MOZ_ASSERT(!(!aStage && aGenerateSpeculativeLoads),
+             "Must not generate speculative loads without a stage");
   MOZ_COUNT_CTOR(nsHtml5TreeBuilder);
 }
 
@@ -167,7 +172,7 @@
 
   // Start wall of code for speculative loading and line numbers
 
-  if (mSpeculativeLoadStage && mode != IN_TEMPLATE) {
+  if (mGenerateSpeculativeLoads && mode != IN_TEMPLATE) {
     switch (aNamespace) {
       case kNameSpaceID_XHTML:
         if (nsGkAtoms::img == aName) {
@@ -986,10 +991,7 @@
     }
     return;
   }
-  if (mSpeculativeLoadStage && aName == nsGkAtoms::picture) {
-    // mSpeculativeLoadStage is non-null only in the off-the-main-thread
-    // tree builder, which handles the network stream
-    //
+  if (mGenerateSpeculativeLoads && aName == nsGkAtoms::picture) {
     // See comments in nsHtml5SpeculativeLoad.h about <picture> preloading
     mSpeculativeLoadQueue.AppendElement()->InitOpenPicture();
     return;
@@ -1117,10 +1119,7 @@
     return;
   }
 
-  if (mSpeculativeLoadStage && aName == nsGkAtoms::picture) {
-    // mSpeculativeLoadStage is non-null only in the off-the-main-thread
-    // tree builder, which handles the network stream
-    //
+  if (mGenerateSpeculativeLoads && aName == nsGkAtoms::picture) {
     // See comments in nsHtml5SpeculativeLoad.h about <picture> preloading
     mSpeculativeLoadQueue.AppendElement()->InitEndPicture();
     return;
@@ -1251,16 +1250,26 @@
 }
 
 void nsHtml5TreeBuilder::SetDocumentCharset(NotNull<const Encoding*> aEncoding,
-                                            int32_t aCharsetSource) {
-  if (mBuilder) {
-    mBuilder->SetDocumentCharsetAndSource(aEncoding, aCharsetSource);
-  } else if (mSpeculativeLoadStage) {
-    mSpeculativeLoadQueue.AppendElement()->InitSetDocumentCharset(
-        aEncoding, aCharsetSource);
-  } else {
-    opSetDocumentCharset opearation(aEncoding, aCharsetSource);
-    mOpQueue.AppendElement()->Init(mozilla::AsVariant(opearation));
+                                            nsCharsetSource aCharsetSource,
+                                            bool aCommitEncodingSpeculation) {
+  MOZ_ASSERT(!mBuilder, "How did we call this with builder?");
+  MOZ_ASSERT(mSpeculativeLoadStage,
+             "How did we call this without a speculative load stage?");
+  mSpeculativeLoadQueue.AppendElement()->InitSetDocumentCharset(
+      aEncoding, aCharsetSource, aCommitEncodingSpeculation);
+}
+
+void nsHtml5TreeBuilder::UpdateCharsetSource(nsCharsetSource aCharsetSource) {
+  MOZ_ASSERT(!mBuilder, "How did we call this with builder?");
+  MOZ_ASSERT(mSpeculativeLoadStage,
+             "How did we call this without a speculative load stage (even "
+             "though we don't need it right here)?");
+  if (mViewSource) {
+    mViewSource->UpdateCharsetSource(aCharsetSource);
+    return;
   }
+  opUpdateCharsetSource operation(aCharsetSource);
+  mOpQueue.AppendElement()->Init(mozilla::AsVariant(operation));
 }
 
 void nsHtml5TreeBuilder::StreamEnded() {